runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2692,21 +2689,40 @@ J9::Z::TreeEvaluator::inlineStringHashCode(TR::Node* node, TR::CodeGenerator* cg
    generateS390LabelInstruction(cg, TR::InstOpCode::label, node, cFlowRegionStart);
    cFlowRegionStart->setStartInternalControlFlow();
 
-   if(!isCompressed)
+   switch (elementType)
       {
-      // registerIndex *= 2 and registerCount *= 2
-      generateRSInstruction(cg, TR::InstOpCode::getShiftLeftLogicalSingleOpCode(), node, registerIndex, registerIndex, 1);
-      generateRSInstruction(cg, TR::InstOpCode::getShiftLeftLogicalSingleOpCode(), node, registerCount, registerCount, 1);
+      case TR::Int8:
+         break;
+      case TR::Int16:
+         // registerIndex *= 2 and registerCount *= 2
+         generateRSInstruction(cg, TR::InstOpCode::getShiftLeftLogicalSingleOpCode(), node, registerIndex, registerIndex, 1);
+         generateRSInstruction(cg, TR::InstOpCode::getShiftLeftLogicalSingleOpCode(), node, registerCount, registerCount, 1);
+         break;
+      case TR::Int32:
+         // registerIndex *= 4 and registerCount *= 4
+         generateRSInstruction(cg, TR::InstOpCode::getShiftLeftLogicalSingleOpCode(), node, registerIndex, registerIndex, 2);
+         generateRSInstruction(cg, TR::InstOpCode::getShiftLeftLogicalSingleOpCode(), node, registerCount, registerCount, 2);
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported vectorizedHashCode element type");
       }
 
    // registerEnd = registerIndex + registerCount
    generateRXInstruction(cg, TR::InstOpCode::getLoadAddressOpCode(), node, registerEnd, generateS390MemoryReference(registerIndex, registerCount, 0, cg));
 
-   // registerHash = 0
-   generateRREInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, registerHash, registerHash);
+   TR::Register* registerHash;
+   if(nonZeroInitial)
+      {
+      registerHash = cg->gprClobberEvaluate(nodeHash);
+      }
+   else
+      {
+      registerHash = cg->allocateRegister();
+      generateRREInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, registerHash, registerHash);
+      }
 
-   // Branch to labelSerial if registerCount < stringSize
-   generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::getCmpLogicalOpCode(), node, registerCount, static_cast<int32_t>(stringSize), TR::InstOpCode::COND_MASK4, labelSerial, false, false);
+   // Branch to labelSerial if registerCount < chunkSize
+   generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::getCmpLogicalOpCode(), node, registerCount, chunkSize, TR::InstOpCode::COND_MASK4, labelSerial, false, false);
```

Comment:
Optional Suggetion: As we are here, to increase the readability of the code, may be we can use `COND_BL` instead of `COND_MASK4`